doc/developer-guide/internal-libraries/Errata.en.rst

```diff
@@ -0,0 +1,133 @@
+.. Licensed to the Apache Software Foundation (ASF) under one
+   or more contributor license agreements. See the NOTICE file distributed with this work for
+   additional information regarding copyright ownership. The ASF licenses this file to you under the
+   Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+   the License. You may obtain a copy of the License at
+
+   http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software distributed under the License
+   is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+   or implied. See the License for the specific language governing permissions and limitations under
+   the License.
+
+.. include:: ../../common.defs
+
+.. default-domain:: cpp
+
+Errata
+*******
+
+Synopsis
+========
+
+:code:`#include <tsconfig/Errata.h>`
+
+:class:`Errata` is an error reporting class. It is designed to make it easier to report errors
+to callers. It provides a stack of error messages, called *annotations*, each with a severity.
+:class:`Errata` is optimized for the success (no annotations) case, which is the common case. It uses
+an internal shared pointer architecture so returning instances from functions is fast (no copies).
+The annotations, or "notes", are allocated in a :class:`MemArena` to minimize memory allocation costs.
```

Comment:
Would the use of this ever be high-frequency enough to be worth it to optimize its use?

Is it used for "assert" type situations?  I think we should avoid any heap allocation in such cases if possible.  Could we use FixedBufferWriter into a large statically-allocated array, and output it using a write() syscall?

Reply:
Use of what? `Errata`? Yes, I would very much like to make it heavily used. The success case? Hopefully the rest of our code does in fact work almost all the time.

I have considered the static array issue. The problem with that is then you have the static array everywhere, and you've bulked up the by far more common case to handle an exceptional case. I think that if the static allocation was a serious issue, you'd do better to do your latter suggestion and use `FixedBufferWriter` into a static array. I'd probably handle that in a TS oriented exception class, which lets the call site decide if memory allocation is contra-indicated or not.